angular2-jwt.ts

```diff
@@ -56,57 +64,74 @@ export class AuthHttp {
 
   private _config: IAuthConfig;
   public tokenStream: Observable<string>;
+  private jwtHelper:JwtHelper;
 
   constructor(options: AuthConfig, private http: Http) {
     this._config = options.getConfig();
+    this.jwtHelper = new JwtHelper();
 
     this.tokenStream = new Observable((obs: any) => {
       obs.next(this._config.tokenGetter())
     });
   }
 
-  _request(url: string | Request, options?: RequestOptionsArgs) : Observable<Response> {
+  _request(req: Request) : Observable<Response> {
 
     let request:any;
-    
-    if(!tokenNotExpired(null, this._config.tokenGetter())) {
-      if(!this._config.noJwtError) {
-        throw 'Invalid JWT';
+
+    // this is observable as it may need to come from an http call ...
+    let refreshToken:Observable<string>;
+
+    var token = this._config.tokenGetter()
+    // do we even have a token?
+    if (token === null) {
+      // no, is this an error?
+      if(this._config.noJwtError) {
+        // no, we can provide this token to the request
+        refreshToken = Observable.of(token)
       } else {
-        request = this.http.request(url, options);
-      }
-      
-    } else if(typeof url === 'string') {
-      let reqOpts = options || {};
-      
-      if(!reqOpts.headers) {
-        reqOpts.headers = new Headers();
+        // yes, we can't do anything
+        throw 'Invalid JWT';
       }
-      
-      reqOpts.headers.set(this._config.headerName, this._config.headerPrefix + this._config.tokenGetter());
-      request = this.http.request(url, reqOpts);
-      
-    } else {
-      let req:Request = <Request>url;
-      
-      if(!req.headers) {
-        req.headers = new Headers();
+    } else{
+      // yes, we have a token ... is it still valid? (with some allowance)
+      if (this.jwtHelper.isTokenExpired(token, this._config.refreshOffset)) {
+        // no, it's expired - we'll call the token refresh prior to the call
+        refreshToken = this._config.refresh();
+      } else {
+        // yes, it's still good so we can provide this token to the request
+        refreshToken = Observable.of(token);
       }
-      
-      req.headers.set(this._config.headerName, this._config.headerPrefix + this._config.tokenGetter());
-      request = this.http.request(req);
     }
-    
+
+    if(!req.headers) {
+      req.headers = new Headers();
+    }
+
+    // chain it after the refresh call
+    request = refreshToken.flatMap(token => {
+      if (token === null) {
```

Comment:
The token can be null - if someone isn't logged in then the token getter would already return null so this extended that to tokens expiring and not being refreshed.